ow_power_system/src/power_system_node.cpp

```diff
@@ -17,27 +25,117 @@ int main(int argc, char* argv[]) {
 
   //Construct our power draw listener
   //TODO: Check out latching
-  ros::Subscriber power_listener = n.subscribe("power_draw", 1000, computeSOC);
+  //ros::Subscriber power_listener = n.subscribe("power_draw", 1000, computeSOC);
 
-  //Construct our State of Charge (SOC) publisher
+  //Construct our State of Charge (SOC) publisher (Volts)
   ros::Publisher SOC_pub = n.advertise<std_msgs::Float64>("state_of_charge",1000);
+  //Construct our Remaining Useful Life (RUL) publisher (Seconds)
+  ros::Publisher RUL_pub = n.advertise<std_msgs::Float64>("remaining_useful_life",1000);
+  
+  //Load power values csv
+  string csv_path,csv_file;
+  csv_file = "/data/onewatt.csv";
+  //FILE OPTIONS: onewatt.csv, eightwatt.csv, sixteenwatt.csv
+  csv_path = ros::package::getPath("ow_power_system");
+  csv_path += csv_file;
+
+  auto power_csv = loadCSV(csv_path);
+  
+  //Define our publication rate and initialize our time to 0
+  double power_update_rate;  
+  if (csv_file=="/data/onewatt.csv"){
+    power_update_rate = 0.10; //csv is at 10Hz intervals
+  } else {
+    power_update_rate = 1.0; // 1 Hz default
+  }
+  int t_step= 1;//starting at 1 not zero to skip csv headings
 
+  //set our indices
+  int time_i = 0;
+  int voltage_i = 1;
+  int rul_i = 2;
+  
   // ROS Loop. Note that once this loop starts,
   // this function (and node) is terminated with an interrupt.
-  ros::Rate rate(1); // 1 Hz, tailor as needed
-  while (ros::ok()) {
-    //individual soc_msg to be published by SOC_pub
-    std_msgs::Float64 soc_msg;
+  
+  ros::Rate rate(power_update_rate);
+  //individual soc_msg to be published by SOC_pub
+  std_msgs::Float64 soc_msg;
+  std_msgs::Float64 rul_msg;
+  ROS_INFO ("Power system node running");
 
-    soc_msg.data=0.0; //zero for testing purposes, will be calculated later
+  int power_lifetime = power_csv.size();
+  while (ros::ok()) {
+    if (t_step < power_lifetime) {
+      soc_msg.data=power_csv[t_step][voltage_i];
+      rul_msg.data=power_csv[t_step][rul_i];
+    } //should keep publishing last voltage value if sim time exceeds the length of the csv
+    
+    //soc_msg.data=t;
+    //soc_msg.data=getSampleValue(timestamp, power_csv); //zero for testing purposes, will be calculated later
 
     //publish current SOC
     SOC_pub.publish(soc_msg);
+    RUL_pub.publish(rul_msg);
+    t_step+=1;
     
     ros::spinOnce();
     rate.sleep();
-    ROS_INFO ("Power system node running");
+   
   }
 
   return 0;
 }
+
+
+/* Function to load pre-generated values as a vector of vectors of doubles
+ * Returns std::vector<std::vector<std::double>>
+ */
+vector<vector<double>> loadCSV(const std::string& filename){
+  ifstream power_csv(filename);
+  if (power_csv.fail()) {
+    ROS_WARN("Loading power csv file has failed. power_system_node will publish zeros.");
+    vector<double> zero_row = {0.0,0.0,0.0};
+    vector<vector<double>> zeros;
+    zeros.push_back(zero_row);
+    zeros.push_back(zero_row);
```

Comment:
there reason you are adding two **zero_row** here is because **t_step** needlessly set to 1.

Reply:
For all other csvs, there is a header row to skip -- I'll change it in the loadCSV function so I can set t_step to 0 and remove the second zero_row.

Reply:
Actually, wait, I was looking at my column loading. I'd just have to repeat a line similarly in the section that actually matters, so I'd prefer to keep it the way it is. It's meant to be a temporary solution anyway, so it shouldn't really matter - do you feel strongly that it has to be done exactly this way?

Reply:
Nope, it is not a big deal. I just saw that your loadCSV function skips the csv header line and then adds samples.

Reply:
Ah thanks, I'm with you now - just fixed it.